packages/SwingSet/docs/virtual-objects.md

```diff
@@ -0,0 +1,111 @@
+
+A vat can define new kinds of virtual object by calling the `makeKind` function provided as a vat global:
+
+`makeKind(bodyMaker)`
+
+The return value from `makeKind` is a maker function which the vat can use to create instances of the newly defined object kind.
+
+The `bodyMaker` parameter is a user provided function that binds an in-memory instance of the virtual object with the virtual object's state, associating such instances with the virtual object's behavior.  It is called whenever a new virtual object instance is created, whenever such an instance is swapped in from secondary storage, and whenever a reference to a virtual object is received as a parameter of a message and deserialized.
+
+The `bodyMaker` function is passed the virtual object's state as a parameter, and is expected to return a new Javascript object with methods that close over this state.
+
+One distinguished method, named `initialize`, is called when new instances are first created, and is expected to initialize the state object.  Any parameters passed to the maker function returned by `makeKind` are passed directly to the `initialize` method.
+
+For example:
+
+```javascript
+function counterBody(state) {
+  return {
+    initialize(name = 'thing') {
+      state.counter = 0;
+      state.name = name;
+    },
+    inc() {
+      state.counter += 1;
+    },
+    dec() {
+      state.counter -= 1;
+    },
+    reset() {
+      state.counter = 0;
+    },
+    rename(newName) {
+      state.name = newName;
+    },
+    getCount() {
+      return state.counter;
+    },
+    getName() {
+      return state.name;
+    },
+  };
+}
+```
+
+This function defines the body for a simple counter object with two properties in its state, a count and a name.
+
+You'd define the corresponding virtual object kind with:
+
+`const counterMaker = makeKind(counterBody);`
+
+and then use it like this:
+
+```javascript
+const fooCounter = counterMaker('foo');
+const barCounter = counterMaker('bar');
+
+fooCounter.inc();
+fooCounter.inc();
+barCounter.inc();
+barCounter.rename('new bar');
+console.log(`${fooCounter.getName()} count is ${fooCounter.getCount()`); // "foo count is 2"
+console.log(`${barCounter.getName()} count is ${barCounter.getCount()`); // "new bar count is 1"
+```
+
+Additional important details:
+
+- The set of state properties is determined by the `initialize` method.  The properties that exist when `initialize` returns _are_ the properties of the object.  State properties cannot thereafter be added or removed.
+
+- The values a state property may take are limited to things that are serializable and which may be hardened and made immutable (and, in fact, _are_ hardened and made immutable the moment they are assigned).  That is, you can replace what value a state property _has_, but you cannot modify a state property's value in place.  In other words, you can do things like:
```

Comment:
```suggestion
- The values a state property may take are limited to things that are serializable and which may be hardened (and, in fact, _are_ hardened, serialized, and unserialized the moment they are assigned).  That is, you can replace what value a state property _has_, but you cannot modify a state property's value in place.  In other words, you can do things like:
```

Because it has, as you say below, "references to other virtual objects", it is ***not*** immutable.

Reply:
Well, it _is_ immutable, but it's shallowly immutable rather than transitively immutable.  Or are we using "immutable" in our lexicon to always mean transitively immutable?

Reply:
Anyway, I've adopted your wording.